src/python/dxpy/app_builder.py

```diff
@@ -150,20 +203,54 @@ def upload_resources(src_dir, project=None, folder='/'):
 
     resources_dir = os.path.join(src_dir, "resources")
     if os.path.exists(resources_dir) and len(os.listdir(resources_dir)) > 0:
-        logger.debug("Uploading in " + src_dir)
-
-        with tempfile.NamedTemporaryFile(suffix=".tar.gz") as tar_fh:
-            subprocess.check_call(['tar', '-C', resources_dir, '-czf', tar_fh.name, '.'])
-            if 'folder' in applet_spec:
-                try:
-                    dxpy.get_handler(dest_project).new_folder(applet_spec['folder'], parents=True)
-                except dxpy.exceptions.DXAPIError:
-                    pass # TODO: make this better
-            target_folder = applet_spec['folder'] if 'folder' in applet_spec else folder
-            dx_resource_archive = dxpy.upload_local_file(tar_fh.name, wait_on_close=True,
-                                                         project=dest_project, folder=target_folder, hidden=True)
-            archive_link = dxpy.dxlink(dx_resource_archive.get_id())
-            return [{'name': 'resources.tar.gz', 'id': archive_link}]
+        target_folder = applet_spec['folder'] if 'folder' in applet_spec else folder
+
+        # Optimistically look for a resource bundle with the same
+        # contents, and reuse it if possible. The resource bundle
+        # carries a property 'resource_bundle_checksum' that indicates
+        # the checksum; the way in which the checksum is computed is
+        # given in the documentation of _directory_checksum.
+        directory_checksum = _directory_checksum(resources_dir)
+        existing_resources = dxpy.find_one_data_object(
+            project=dest_project,
+            folder=target_folder,
+            properties=dict(resource_bundle_checksum=directory_checksum),
+            visibility='either',
+            zero_ok=True,
+            return_handler=True
+        )
+
+        if existing_resources:
+            logger.debug("Found existing resource bundle that matches local resources directory: " +
```

Comment:
We now have logging configured to print only INFO and higher priority messages, so for me, these messages don't show up with any of `dx build -a`, `_DX_DEBUG=1 dx build -a`, or `_DX_DEBUG=2 dx build -a`. Is that intentional? It may be best to also have these messages use logger.info, and also to set logging level to debug if `_DX_DEBUG` is set (though that's outside the scope of this PR).